sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java

```diff
@@ -76,12 +77,13 @@ public void onNext(T value) {
         // Record the initial phase in case we are in the inbound gRPC thread where the phase won't
         // advance.
         int initialPhase = phase;
-        while (!outboundObserver.isReady()) {
+        // A negative phase indicates that the phaser is terminated.
+        while (phase >= 0 && !outboundObserver.isReady()) {
           try {
             phase = phaser.awaitAdvanceInterruptibly(phase, waitTime, TimeUnit.SECONDS);
           } catch (TimeoutException e) {
             totalTimeWaited += waitTime;
-            waitTime = waitTime * 2;
+            waitTime = Math.min(waitTime * 2, 60);
```

Comment:
there is a hard coded constant here (60 s) possibly add some comments for its purpose?